### PR TITLE
[tune] lazy trials

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Create Trial objects one by one from search algorithms. This way the scheduled trials get naturally limited by concurrency limits of the searcher (through `ConcurrencyLimiter`) and the `TrialRunner` (through resources).

## Related issue number

#10138

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by krfricke** [comment]
Open question: Should we remove the `next_trials()` method completely in favor for `next_trial()`?

File: python/ray/tune/suggest/basic_variant.py [hunk 1]
```diff
@@ -74,15 +68,25 @@ def add_configurations(self, experiments):
...
+        except StopIteration:
+            self.set_finished()
+            return None
+
     def next_trials(self):
         """Provides Trial objects to be queued into the TrialRunner.
 
         Returns:
-            trials (list): Returns a list of trials.
+            List[Trial]: A list of trials for the Runner to consume.
```
Comment (richardliaw) [inline python/ray/tune/suggest/basic_variant.py:87]: yeah let's remove

File: python/ray/tune/suggest/search.py [hunk 2]
```diff
@@ -46,11 +45,11 @@ def add_configurations(
         """
         raise NotImplementedError
 
-    def next_trials(self) -> List[Trial]:
-        """Provides Trial objects to be queued into the TrialRunner.
+    def next_trial(self):
```
Comment (richardliaw) [inline python/ray/tune/suggest/search.py:48]:
```suggestion
    def next_trial(self) -> Trial:
```

File: python/ray/tune/tests/test_progress_reporter.py [hunk 3]
```diff
@@ -73,32 +73,12 @@ def f(config):
...
-| f_xxxxx_00014 | PENDING  |       |     |   4 |
-| f_xxxxx_00015 | PENDING  |       |     |   5 |
-| f_xxxxx_00016 | PENDING  |       |     |   6 |
-| f_xxxxx_00017 | PENDING  |       |     |   7 |
-| f_xxxxx_00018 | PENDING  |       |     |   8 |
-| f_xxxxx_00019 | PENDING  |       |     |   9 |
-| f_xxxxx_00000 | RUNNING  |       |   0 |     |
-+---------------+----------+-------+-----+-----+
-... 10 more trials not shown (10 PENDING)"""
+EXPECTED_END_TO_END_START = """Number of trials: 1 (1 RUNNING)
```
Comment (richardliaw) [inline python/ray/tune/tests/test_progress_reporter.py:76]: I think in this PR, let's be sure to show `Total`.

**Comment by krfricke** [comment]
I added the total number of samples to the printout. Also implemented (virtually) infinite samples when `num_samples=-1`

**Review by richardliaw — APPROVED** [review]
nice! the output looks great too.

**Pull request merged**